File: src/shared/schemas/candidate.py
"""Pydantic models for Candidate (SPEC-0A.2).

Candidate is the aesthetic-choice payload emitted by phase agents
(P4 TTS / P5 BGM / P7 Storyboard / P8 Keyframe / P9 Cover) and rendered
by the frontend `CandidateSelector`. Each phase may emit at most three
candidates; the container `CandidateList` enforces that upper bound.
"""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, Tuple

from pydantic import BaseModel, ConfigDict, Field

PreviewType = Literal["audio", "image", "video", "color_palette"]

PREVIEW_TYPES: Tuple[PreviewType, ...] = (
    "audio",
    "image",
    "video",
    "color_palette",
)


class _Strict(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Candidate(_Strict):
    candidate_id: str = Field(pattern=r"^cand_[A-Za-z0-9_-]+$")
    preview_url: str = Field(min_length=1)
    preview_type: PreviewType
    style_tags: List[str]
    description: str = Field(min_length=1)
    is_recommended: bool
    adjustable_params: Dict[str, Any]
    rationale: str = Field(min_length=1)
    raw_bgm_url: Optional[str] = None


class CandidateList(_Strict):
    """Container enforcing the SPEC-0A.2 rule that candidates count <= 3."""

    candidates: List[Candidate] = Field(max_length=3)


__all__ = [
    "PREVIEW_TYPES",
    "Candidate",
    "CandidateList",
    "PreviewType",
]

File: src/shared/types/candidate.ts
// Candidate type (SPEC-0A.2). Mirror of src/shared/schemas/candidate.py.

export type PreviewType = "audio" | "image" | "video" | "color_palette";

export interface Candidate {
  candidate_id: string;            // 'cand_' + UUID
  preview_url: string;             // preview asset URL (trimmed clip)
  preview_type: PreviewType;
  style_tags: string[];
  description: string;
  is_recommended: boolean;         // system recommendation flag
  adjustable_params: Record<string, unknown>;
  rationale: string;               // why this was recommended
  raw_bgm_url?: string;            // full-length BGM (P5 only)
}

// Container enforcing the SPEC-0A.2 cap of 3 candidates per phase.
export interface CandidateList {
  candidates: Candidate[];
}
